### ISSUE TITLE
App Instrumented w/ AI suddenly throwing a 'Method not found' exception

### ISSUE THREAD

**Issue by abaranch** [Microsoft/ApplicationInsights-dotnet-server#147]
Copied from https://social.msdn.microsoft.com/Forums/vstudio/en-US/e76b881f-3f2d-4182-8b2f-fe2654c4c0ce/app-instrumented-w-ai-suddenly-throwing-a-method-not-found-exception?forum=ApplicationInsights

Out of the blue, an app we have instrumented w/ AI started throwing an exception when attempting to connect to our DB. We can't seem to identify why. Any ideas? Below is the YSOD details we're seeing. 

```
Method not found: 'System.Object System.Diagnostics.DebuggerHiddenAttribute.ApplicationInsights_OnBegin(Int32, System.Object, System.Object, System.Object)'.
[MissingMethodException: Method not found: 'System.Object System.Diagnostics.DebuggerHiddenAttribute.ApplicationInsights_OnBegin(Int32, System.Object, System.Object, System.Object)'.]
   System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method) +0
   System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior) +37
   System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior) +12
   Glimpse.Ado.AlternateType.GlimpseDbCommand.ExecuteDbDataReader(CommandBehavior behavior) +225
   System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior) +12
   System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.<Reader>b__c(DbCommand t, DbCommandInterceptionContext`1 c) +14
   System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch(TTarget target, Func`3 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed) +72
   System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.Reader(DbCommand command, DbCommandInterceptionContext interceptionContext) +355
   System.Data.Entity.Internal.InterceptableDbCommand.ExecuteDbDataReader(CommandBehavior behavior) +166
   System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior) +12
   System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior) +36

[EntityCommandExecutionException: An error occurred while executing the command definition. See the inner exception for details.]
   System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior) +111
   System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlan.Execute(ObjectContext context, ObjectParameterCollection parameterValues) +740
   System.Data.Entity.Core.Objects.<>c__DisplayClass7.<GetResults>b__6() +97
   System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction(Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess) +288
   System.Data.Entity.Core.Objects.<>c__DisplayClass7.<GetResults>b__5() +154
   System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute(Func`1 operation) +190
   System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults(Nullable`1 forMergeOption) +279
   System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__0() +16
   System.Data.Entity.Internal.LazyEnumerator`1.MoveNext() +45
   System.Linq.Enumerable.FirstOrDefault(IEnumerable`1 source) +183
   System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.<GetElementFunction>b__1(IEnumerable`1 sequence) +40
   System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.ExecuteSingle(IEnumerable`1 query, Expression queryRoot) +60
   System.Data.Entity.Core.Objects.ELinq.ObjectQueryProvider.System.Linq.IQueryProvider.Execute(Expression expression) +113
   System.Data.Entity.Internal.Linq.DbQueryProvider.Execute(Expression expression) +113
   System.Linq.Queryable.FirstOrDefault(IQueryable`1 source) +211
   ShiftDigital.Reach.Core.Auth.UserStore`1.FindByNameAsync(String userName) in C:\shiftdigitalgit\Fiat - Europe (Git)\Reach3-Main\Reach.Core\Auth\UserStore.cs:120
   Microsoft.AspNet.Identity.UserManager`2.FindByNameAsync(String userName) +56
   Microsoft.AspNet.Identity.<>c__DisplayClassa`2.<FindByName>b__9() +20
   System.Threading.Tasks.Task`1.InnerInvoke() +56
   System.Threading.Tasks.Task.Execute() +61
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +135
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +64
   Microsoft.AspNet.Identity.AsyncHelper.RunSync(Func`1 func) +137
   Microsoft.AspNet.Identity.UserManagerExtensions.FindByName(UserManager`2 manager, String userName) +157
   ShiftDigital.Reach.Core.Auth.UserManager.Create(IdentityFactoryOptions`1 options, IOwinContext context) in C:\shiftdigitalgit\Fiat - Europe (Git)\Reach3-Main\Reach.Core\Auth\UserManager.cs:24
   Microsoft.AspNet.Identity.Owin.IdentityFactoryProvider`1.Create(IdentityFactoryOptions`1 options, IOwinContext context) +19
   Microsoft.AspNet.Identity.Owin.<Invoke>d__0.MoveNext() +104
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.StageAsyncResult.End(IAsyncResult ar) +70
   Microsoft.Owin.Host.SystemWeb.IntegratedPipeline.IntegratedPipelineContext.EndFinalWork(IAsyncResult ar) +64
   System.Web.AsyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +484
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +157
```

**Comment by SergeyKanzhelev** [comment]
There is more fresh report of the same (at least from the looks of it) issue: https://social.msdn.microsoft.com/Forums/vstudio/en-US/576ae696-f41c-4081-a875-27c615bd73fb/ai-on-azure-cloud-service-webrole-fails-on-app-pool-restart?forum=ApplicationInsights&prof=required

**Comment by SergeyKanzhelev** [comment]
I still do not know the root cause, but I believe the issue is no longer crashing customer application. Here is my investigation:

`Decorate` call in Intercept NuGet used to work like this. First, there will be an attempt to inject callbacks into the method of interest (call to `DecorateInternal`) and then we will add callbacks into the collection. Note, that exception in Callbacks.Add will be swallowed and never showed up:

``` csharp
var id = Internal.DecoratorInternal.DecorateInternal(assemblyName, moduleName, methodName, argsCount);
try
{
//this is temporary solution before mscorlib injection will be implemented
Callbacks.Add(id, beginCallback, endCallback, exceptionCallback);
}
catch (Exception)
{
}
```

So if RTIA failed to create callbacks (like this `Method not found: System.Object System.Diagnostics.DebuggerHiddenAttribute.ApplicationInsights_OnBegin`) into `mscorlib` - we still inject the calls to this callbacks. 

In the latest version of Intercept the ordering has changed. Now `Callbacks.Add` will be called first and then `DecorateInternal` will be called.

So if RTIA failed to create callbacks - the method `Callbacks.Add` will fail first and will not attempt to decorate the method. I found instances of the corresponding error:

```
AI (Internal): RemoteDependency profiler failed to attach. Collection will default to EventSource implementation. Error details: System.MissingMethodException: Method not found: 'Void System.Diagnostics.DebuggerHiddenAttribute.ApplicationInsights_AddCallbacks(Int32, System.Func`2<System.Object,System.Object>, System.Func`4<System.Object,System.Object,System.Object,System.Object>, System.Action`3<System.Object,System.Exception,System.Object>)'.
   at Microsoft.Diagnostics.Instrumentation.Extensions.Base.Callbacks.Add(Int32 methodId, Func`2 onBegin, Func`4 onEnd, Action`3 onException)
   at Microsoft.Diagnostics.Instrumentation.Extensions.Intercept.Decorator.Decorate(String assemblyName, String moduleName, String methodName, UInt32 argsCount, Func`2 beginCallback, Func`4 endCallback, Action`3 exceptionCallback)
   at Microsoft.Diagnostics.Instrumentation.Extensions.Intercept.Functions.Decorate(String assemblyName, String moduleName, String methodName, Func`2 beginCallback, Func`4 endCallback, Action`3 exceptionCallback, Boolean isStatic, Boolean isSafe)
   at Microsoft.ApplicationInsights.DependencyCollector.Implementation.ProfilerRuntimeInstrumentation.DecorateProfilerForHttp(ProfilerHttpProcessing& httpCallbacks)
   at Microsoft.ApplicationInsights.DependencyCollector.DependencyTrackingTelemetryModule.InitializeForRuntimeProfiler()
   at Microsoft.ApplicationInsights.DependencyCollector.DependencyTrackingTelemetryModule.InitializeForRuntimeInstrumentationOrFramework()
```

So application wouldn't crash, but Application Insights SDK will fall back into framework data collection.

**Comment by SergeyKanzhelev** [comment]
Next steps for me are:
- understand the root cause
- add a safety check that callbacks exists before injecting them into the method
- find the root cause - why callbacks wasn't created in `mscorlib`

**Comment by SergeyKanzhelev** [comment]
The version of Intercept NuGet that fixes the problem is `2.0.5` [here](https://www.nuget.org/packages/Microsoft.ApplicationInsights.Agent.Intercept/2.0.5). 

WAD Agent with the updated Intercept version should be released this week. All the changes are already in.

Closing this issue and will track work to find the root case separately

CC: @Dmitry-Matveev
